### PR TITLE
[FLINK-26823] Adding Cassandra Table Source

### PR DESCRIPTION
**Pull request by Poorvankbhatia**
## Summary
  This PR implements complete **Apache Flink Table/SQL API support for reading from Apache Cassandra** through a new `CassandraDynamicTableSource`. Previously, users could only read from Cassandra using the
  DataStream API with the existing `CassandraSource`. This change enables direct SQL queries against Cassandra tables using Flink's Table API.

  **Key capabilities added:**
  - Full SQL query support: `SELECT * FROM cassandra_table WHERE conditions`
  - Complete type mapping between all Cassandra data types and Flink logical types
  - Support for complex nested structures (UDTs, collections, arrays)

  ## Changes Made

  ###  Core Table API Implementation
  - **`CassandraDynamicTableSource`**: New DynamicTableSource implementation 

  ###  Comprehensive Type Mapping System
  - **`RowToRowDataMapper`**: Main orchestrator for field-level conversions
  - **`CassandraFieldMapperFactory`**: Factory for creating appropriate mappers based on Flink logical types
  - **`PrimitiveFieldMappers`**:
    - Added `DynamicDecimalMapper` for runtime type detection (handles both varint and decimal columns)
    - Added `TimeMapper` for TIME_WITHOUT_TIME_ZONE with precision conversion
    - Enhanced `StringMapper` for UUID/TimeUUID/Duration/Inet types
    - Added `convertValue()` methods to `IntegerMapper` and `LongMapper` to prevent ClassCastException
  - **`CollectionFieldMappers`**:
    - Fixed `SetMapper` count typing for MULTISET support
    - Support for deeply nested structures (List<Row<...>>, Map<String, Row<...>>)

  ###  Build System Enhancements
  - **Enhanced `pom.xml`**:
    - Added comprehensive shaded jar configuration with detailed documentation
    - Added filters to exclude signature files and prevent conflicts
    - Ensures proper DataStax driver dependency resolution in SQL client

  ### Testing
  - **`CassandraTableE2ETest`**: End-to-end tests covering all edge cases
  - **Unit test coverage** for all mapper components with mock-based testing

  ## Type Mapping Details

  | Cassandra Type | Flink LogicalType | Mapper Used | Notes |
  |---|---|---|---|
  | boolean | BOOLEAN | BooleanMapper | Direct mapping |
  | tinyint | TINYINT | ByteMapper | Direct mapping |
  | smallint | SMALLINT | ShortMapper | Direct mapping |
  | int | INTEGER | IntegerMapper | With convertValue() for type safety |
  | bigint | BIGINT | LongMapper | With convertValue() for type safety |
  | float | FLOAT | FloatMapper | Direct mapping |
  | double | DOUBLE | DoubleMapper | Direct mapping |
  | text/ascii | VARCHAR | StringMapper | Direct mapping |
  | uuid/timeuuid | VARCHAR | StringMapper | Converted to string representation |
  | inet | VARCHAR | StringMapper | Converted to IP address string |
  | duration | VARCHAR | StringMapper | Converted to ISO-8601 duration string |
  | decimal | DECIMAL | DynamicDecimalMapper | Runtime type detection |
  | varint | DECIMAL | DynamicDecimalMapper | Runtime type detection |
  | date | DATE | DateMapper | Converted to epoch days |
  | time | TIME_WITHOUT_TIME_ZONE | TimeMapper | Nanoseconds → milliseconds |
  | timestamp | TIMESTAMP_WITHOUT_TIME_ZONE | TimestampMapper | Direct mapping |
  | blob | VARBINARY | BinaryMapper | ByteBuffer → byte[] |
  | list<T> | ARRAY<T> | ArrayMapper | Recursive mapping |
  | map<K,V> | MAP<K,V> | MapMapper | Recursive mapping |
  | set<T> | MULTISET<T> | SetMapper | Converted to Map<element, count> |
  | UDT | ROW<...> | RowMapper | Field-by-field recursive mapping |

  ## Testing

  ### Unit Tests
  - [x] `CassandraFieldMapperFactoryTest`: Tests all mapper creation scenarios
  - [x] `PrimitiveFieldMappersTest`: Tests all primitive type conversions with edge cases
  - [x] `CollectionFieldMappersTest`: Tests nested collection handling
  - [x] `CassandraDynamicTableSourceTest`: Tests table source configuration and split generation

  ### Integration Tests
  - [x] `CassandraDynamicTableSourceITCase`: Comprehensive end-to-end test that verifies:
    - Correct type mapping for all primitive types
    - Nested structure handling (arrays of UDTs, maps with Row values)
    - Null value handling in collections
    - Empty collection behavior
    - Large binary data handling
    - High precision numeric types
    - UUID/TimeUUID string conversion
    - Duration and Inet address handling

  ### Build Verification
  - [x] All tests pass: `mvn clean package`
  - [x] Code formatting: `mvn spotless:check`
  - [x] Checkstyle compliance: `mvn checkstyle:check`

  ## Breaking Changes
None. This is a new feature that adds Table API support while maintaining full backward compatibility with existing DataStream API usage.

### CODE REVIEW

**Comment by Poorvankbhatia** [comment]
@MartijnVisser please take a look. Thanks

**Comment by echauchot** [comment]
> ## Summary
> This PR implements complete **Apache Flink Table/SQL API support for reading from Apache Cassandra** through a new `CassandraDynamicTableSource`. Previously, users could only read from Cassandra using the DataStream API with the existing `CassandraSource`. This change enables direct SQL queries against Cassandra tables using Flink's Table API.
> 
> **Key capabilities added:**
> 
> * Full SQL query support: `SELECT * FROM cassandra_table WHERE conditions`
> * Complete type mapping between all Cassandra data types and Flink logical types
> * Support for complex nested structures (UDTs, collections, arrays)
> 
> ## Changes Made
> ### Core Table API Implementation
> * **`CassandraDynamicTableSource`**: New DynamicTableSource implementation
> 
> ### Comprehensive Type Mapping System
> * **`RowToRowDataMapper`**: Main orchestrator for field-level conversions
> * **`CassandraFieldMapperFactory`**: Factory for creating appropriate mappers based on Flink logical types
> * **`PrimitiveFieldMappers`**:
>   
>   * Added `DynamicDecimalMapper` for runtime type detection (handles both varint and decimal columns)
>   * Added `TimeMapper` for TIME_WITHOUT_TIME_ZONE with precision conversion
>   * Enhanced `StringMapper` for UUID/TimeUUID/Duration/Inet types
>   * Added `convertValue()` methods to `IntegerMapper` and `LongMapper` to prevent ClassCastException
> * **`CollectionFieldMappers`**:
>   
>   * Fixed `SetMapper` count typing for MULTISET support
>   * Support for deeply nested structures (List<Row<...>>, Map<String, Row<...>>)
> 
> ### Build System Enhancements
> * **Enhanced `pom.xml`**:
>   
>   * Added comprehensive shaded jar configuration with detailed documentation
>   * Added filters to exclude signature files and prevent conflicts
>   * Ensures proper DataStax driver dependency resolution in SQL client
> 
> ### Testing
> * **`CassandraTableE2ETest`**: End-to-end tests covering all edge cases
> * **Unit test coverage** for all mapper components with mock-based testing
> 
> ## Type Mapping Details
> Cassandra Type	Flink LogicalType	Mapper Used	Notes
> boolean	BOOLEAN	BooleanMapper	Direct mapping
> tinyint	TINYINT	ByteMapper	Direct mapping
> smallint	SMALLINT	ShortMapper	Direct mapping
> int	INTEGER	IntegerMapper	With convertValue() for type safety
> bigint	BIGINT	LongMapper	With convertValue() for type safety
> float	FLOAT	FloatMapper	Direct mapping
> double	DOUBLE	DoubleMapper	Direct mapping
> text/ascii	VARCHAR	StringMapper	Direct mapping
> uuid/timeuuid	VARCHAR	StringMapper	Converted to string representation
> inet	VARCHAR	StringMapper	Converted to IP address string
> duration	VARCHAR	StringMapper	Converted to ISO-8601 duration string
> decimal	DECIMAL	DynamicDecimalMapper	Runtime type detection
> varint	DECIMAL	DynamicDecimalMapper	Runtime type detection
> date	DATE	DateMapper	Converted to epoch days
> time	TIME_WITHOUT_TIME_ZONE	TimeMapper	Nanoseconds → milliseconds
> timestamp	TIMESTAMP_WITHOUT_TIME_ZONE	TimestampMapper	Direct mapping
> blob	VARBINARY	BinaryMapper	ByteBuffer → byte[]
> list	ARRAY	ArrayMapper	Recursive mapping
> map<K,V>	MAP<K,V>	MapMapper	Recursive mapping
> set	MULTISET	SetMapper	Converted to Map<element, count>
> UDT	ROW<...>	RowMapper	Field-by-field recursive mapping
> ## Testing
> ### Unit Tests
> * [x]  `CassandraFieldMapperFactoryTest`: Tests all mapper creation scenarios
> * [x]  `PrimitiveFieldMappersTest`: Tests all primitive type conversions with edge cases
> * [x]  `CollectionFieldMappersTest`: Tests nested collection handling
> * [x]  `CassandraDynamicTableSourceTest`: Tests table source configuration and split generation
> 
> ### Integration Tests
> * [x]  `CassandraDynamicTableSourceITCase`: Comprehensive end-to-end test that verifies:
>   
>   * Correct type mapping for all primitive types
>   * Nested structure handling (arrays of UDTs, maps with Row values)
>   * Null value handling in collections
>   * Empty collection behavior
>   * Large binary data handling
>   * High precision numeric types
>   * UUID/TimeUUID string conversion
>   * Duration and Inet address handling
> 
> ### Build Verification
> * [x]  All tests pass: `mvn clean package`
> * [x]  Code formatting: `mvn spotless:check`
> * [x]  Checkstyle compliance: `mvn checkstyle:check`
> 
> ## Breaking Changes
> None. This is a new feature that adds Table API support while maintaining full backward compatibility with existing DataStream API usage.

There is actually a breaking change on CassandraSource creation (constructor replaced with builder). But I agree with this change and it is ok as this class is annotated PublicEvolving

**Review by echauchot — COMMENTED** [review]
Thanks for your work @Poorvankbhatia !  This is very good code.
Almost looks good to me. Only minor things.
I reviewed all the production code but not the test code yet. Submitting to give you updates.

**Comment by echauchot** [comment]
please rebase on main as there were merged changes

**Review by echauchot — CHANGES_REQUESTED** [review]
@Poorvankbhatia thanks for your work on tests ! I have some doubts on the unit tests but the integration tests look good modulo some times where we need to test the content of the resultSet and not just isNotNull, isNotEmpty or the content of the inner convert.

File: flink-connector-cassandra/src/main/java/org/apache/flink/connector/cassandra/table/mapper/PrimitiveFieldMappers.java [hunk 1]
```diff
@@ -0,0 +1,375 @@
...
+        @Override
+        public Object convertValue(Object value) {
+            if (value == null) {
+                return null;
+            }
+            // Cassandra time is nanoseconds since midnight (long)
+            // Flink TIME is milliseconds since midnight (int)
+            if (value instanceof Long) {
+                long nanoseconds = (Long) value;
+                return (int) (nanoseconds / 1_000_000); // Convert nanoseconds to milliseconds
```
[the remark this answers is not in the document]
Comment (Poorvankbhatia) [reply flink-connector-cassandra/src/main/java/org/apache/flink/connector/cassandra/table/mapper/PrimitiveFieldMappers.java:233]: I borrowed this idea from JDBC connector : [Link](https://github.com/apache/flink-connector-jdbc/blob/main/flink-connector-jdbc-core/src/main/java/org/apache/flink/connector/jdbc/core/database/dialect/AbstractDialectConverter.java#L167)
So even though Flink’s TimeType logically supports nanosecond precision, but at runtime it's stored as an int representing milliseconds since midnight. Returning a Long causes ClassCastException because  it looks like RowData.getInt() is used for TIME fields.

File: flink-connector-cassandra/src/main/java/org/apache/flink/connector/cassandra/source/CassandraSourceBuilder.java [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+ *
+ * <h3>Usage Examples:</h3>
+ *
+ * <h4>POJO Source (DataStream API):</h4>
+ *
+ * <pre>{@code
+ * CassandraSource<MyPojo> source = CassandraSource.builder()
+ *     .setClusterBuilder(clusterBuilder)
+ *     .setQuery("SELECT * FROM my_keyspace.my_table")
+ *     .setMapperOptions(() -> new Mapper.Option[] {Mapper.Option.consistencyLevel(ANY)})
```
[the remark this answers is not in the document]
Comment (Poorvankbhatia) [reply flink-connector-cassandra/src/main/java/org/apache/flink/connector/cassandra/source/CassandraSourceBuilder.java]: Improved comment.

File: flink-connector-cassandra/src/main/java/org/apache/flink/connector/cassandra/table/mapper/PrimitiveFieldMappers.java [hunk 3]
```diff
@@ -0,0 +1,375 @@
...
+import java.math.BigDecimal;
+import java.math.BigInteger;
+import java.net.InetAddress;
+import java.nio.ByteBuffer;
+import java.time.Instant;
+import java.util.Date;
+
+/** Collection of primitive type field mappers for Cassandra to Flink conversion. */
+@Internal
+public final class PrimitiveFieldMappers {
```
[the remark this answers is not in the document]
Comment (Poorvankbhatia) [reply flink-connector-cassandra/src/main/java/org/apache/flink/connector/cassandra/table/mapper/PrimitiveFieldMappers.java:40]: I think "Internal data structures" refers to Flink's runtime data representations, not that everything needs wrapping.
 - Java primitives (boolean, int, long, etc.) ARE valid internal data structures
  - External types like java.lang.String, java.math.BigDecimal is not - it needs conversion to StringData, DecimalData, etc.
 Let me know if this understanding is correct.

File: flink-connector-cassandra/src/main/java/org/apache/flink/connector/cassandra/table/mapper/PrimitiveFieldMappers.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (echauchot) [reply flink-connector-cassandra/src/main/java/org/apache/flink/connector/cassandra/table/mapper/PrimitiveFieldMappers.java:233]: it seems strange that Flink stores it as an int if it supports nano seconds it should claim to support milliseconds precision as it stores milliseconds since midnight. Maybe something to improve at flink-sql side. 
Any way considering that, fair enough

File: flink-connector-cassandra/src/main/java/org/apache/flink/connector/cassandra/table/mapper/PrimitiveFieldMappers.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (echauchot) [reply flink-connector-cassandra/src/main/java/org/apache/flink/connector/cassandra/table/mapper/PrimitiveFieldMappers.java:40]: I checked in jdbc sql connector AbstractDialectConverter it seems ok, primitive types can be directly set to fields of a GenericRowData object

File: flink-connector-cassandra/src/main/java/org/apache/flink/connector/cassandra/table/CassandraConnectorOptions.java [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+
+package org.apache.flink.connector.cassandra.table;
+
+import org.apache.flink.annotation.PublicEvolving;
+import org.apache.flink.configuration.ConfigOption;
+import org.apache.flink.configuration.ConfigOptions;
+
+/** Options for the Cassandra connector. */
+@PublicEvolving
+public class CassandraConnectorOptions {
```
[the remark this answers is not in the document]
Comment (echauchot) [reply flink-connector-cassandra/src/main/java/org/apache/flink/connector/cassandra/table/CassandraConnectorOptions.java:55]: yes new doc is ok

**Review by etienne-chauchot — COMMENTED** [review]
Thanks @Poorvankbhatia for the fixes. To give you an update: prod code LGTM.
Need to check the test code fixes

File: flink-connector-cassandra/src/test/java/org/apache/flink/connector/cassandra/source/reader/converter/CassandraRowToPojoConverterTest.java [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+
+    @Test
+    void testGetProducedType() {
+        TypeInformation<TestPojo> typeInfo = converter.getProducedType();
+        assertThat(typeInfo).isNotNull();
+        assertThat(typeInfo.getTypeClass()).isEqualTo(TestPojo.class);
+    }
+
+    @Test
+    void testConvertSuccessfully() throws Exception {
```
[the remark this answers is not in the document]
Comment (echauchot) [reply flink-connector-cassandra/src/test/java/org/apache/flink/connector/cassandra/source/reader/converter/CassandraRowToPojoConverterTest.java]: I guess we can leave the unused getters and setters for the pojo fields in case new tests access these fields

File: flink-connector-cassandra/src/test/java/org/apache/flink/connector/cassandra/source/reader/converter/CassandraRowToPojoConverterTest.java [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+import org.mockito.MockitoAnnotations;
+
+import java.util.Date;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+import static org.mockito.Mockito.when;
+
+/** Unit tests for {@link CassandraRowToPojoConverter}. */
+class CassandraRowToPojoConverterTest {
```
[the remark this answers is not in the document]
Comment (echauchot) [reply flink-connector-cassandra/src/test/java/org/apache/flink/connector/cassandra/source/reader/converter/CassandraRowToPojoConverterTest.java:49]: I guess we can leave the unused getters and setters for the pojo fields in case new tests access these fields

**Comment by echauchot** [comment]
> No worries. I'll manage, it would just be longer. It is just a hint to help for future PRs

No worries. I'll manage, it would just be longer. It is just a hint to help for future PRs

File: flink-connector-cassandra/src/test/java/org/apache/flink/connector/cassandra/table/mapper/CollectionFieldMappersTest.java [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Mockito.when;
+
+/** Unit tests for {@link CollectionFieldMappers}. */
+class CollectionFieldMappersTest {
```
[the remark this answers is not in the document]
Comment (echauchot) [reply flink-connector-cassandra/src/test/java/org/apache/flink/connector/cassandra/table/mapper/CollectionFieldMappersTest.java:48]: you missed that one

File: flink-connector-cassandra/src/test/java/org/apache/flink/connector/cassandra/table/CassandraDynamicTableSourceITCase.java [hunk 8]
```diff
@@ -0,0 +1,1520 @@
...
+        assertThat(row.getField(9)).isNotNull();
+        /** map_text_boolean */
+
+        /** Test nested collections */
+        assertThat(row.getField(10)).isNotNull();
+        /** list_of_list */
+        assertThat(row.getField(11)).isNotNull();
+        /** map_of_list */
+        assertThat(row.getField(12)).isNotNull();
+        /** set_of_map */
```
[the remark this answers is not in the document]
Comment (echauchot) [reply flink-connector-cassandra/src/test/java/org/apache/flink/connector/cassandra/table/CassandraDynamicTableSourceITCase.java]: you missed that one

File: flink-connector-cassandra/src/test/java/org/apache/flink/connector/cassandra/table/CassandraDynamicTableSourceITCase.java [hunk 9]
```diff
@@ -0,0 +1,1520 @@
...
+        assertThat(populatedRow.getField(6)).isNotNull();
+        /** large_binary_data */
+        assertThat(populatedRow.getField(7)).isNotNull();
+        /** very_large_binary */
+        assertThat(populatedRow.getField(8)).isEqualTo("populated text");
+        /** empty_text */
+        assertThat(populatedRow.getField(9)).isEqualTo(42);
+        /** zero_int */
+        assertThat(populatedRow.getField(10)).isEqualTo(true);
+        /** false_boolean */
```
[the remark this answers is not in the document]
Comment (echauchot) [reply flink-connector-cassandra/src/test/java/org/apache/flink/connector/cassandra/table/CassandraDynamicTableSourceITCase.java:1445]: you missed that one

File: flink-connector-cassandra/src/test/java/org/apache/flink/connector/cassandra/table/CassandraDynamicTableSourceITCase.java [hunk 10]
```diff
@@ -0,0 +1,1520 @@
...
+
+        Row company = (Row) employeeData.getField(4);
+        assertThat(company.getField(0)).isEqualTo("TechCorp");
+        /** name */
+
+        /** Test collections of deeply nested structures */
+        Object[] employeeList = (Object[]) row.getField(2);
+        assertThat(employeeList).isNotNull().isNotEmpty();
+
+        assertThat(row.getField(3)).isNotNull();
```
[the remark this answers is not in the document]
Comment (echauchot) [reply flink-connector-cassandra/src/test/java/org/apache/flink/connector/cassandra/table/CassandraDynamicTableSourceITCase.java]: assert on the content of the map

File: flink-connector-cassandra/src/test/java/org/apache/flink/connector/cassandra/table/CassandraDynamicTableSourceITCase.java [hunk 11]
```diff
@@ -0,0 +1,1520 @@
...
+        assertThat(row.getField(2)).isNull();
+        /** empty_set */
+        assertThat(row.getField(3)).isNull();
+        /** empty_map */
+
+        /** Test collections with values */
+        Object[] listWithValues = (Object[]) row.getField(4);
+        assertThat(listWithValues).containsExactly("value1", "value2", "value3");
+
+        assertThat(row.getField(5)).isNotNull();
```
[the remark this answers is not in the document]
Comment (echauchot) [reply flink-connector-cassandra/src/test/java/org/apache/flink/connector/cassandra/table/CassandraDynamicTableSourceITCase.java:1213]: you missed that one

File: flink-connector-cassandra/src/test/java/org/apache/flink/connector/cassandra/table/CassandraDynamicTableSourceITCase.java [hunk 12]
```diff
@@ -0,0 +1,1520 @@
...
+
+        /** Test large binary data */
+        assertThat(row.getField(11)).isNotNull();
+        /** large_blob */
+
+        /** Test tuple and UDT with nulls */
+        assertThat(row.getField(12)).isNotNull();
+        /** tuple_with_nulls */
+        assertThat(row.getField(13)).isNotNull();
+        /** partial_udt */
```
[the remark this answers is not in the document]
Comment (echauchot) [reply flink-connector-cassandra/src/test/java/org/apache/flink/connector/cassandra/table/CassandraDynamicTableSourceITCase.java]: missed that one

**Review by echauchot — CHANGES_REQUESTED** [review]
Thanks @Poorvankbhatia for addressing the comments:
- prod code LGTM
- in test code almost LGTM: you missed some of the comments, I went through all the comments and reopened missed ones. 
- Also I have a general comment to test the content of the fields each time you test isNotNull() (nullity check is not enough)

File: flink-connector-cassandra/src/test/java/org/apache/flink/connector/cassandra/source/reader/CassandraRowEmitterTest.java [hunk 13]
```diff
@@ -0,0 +1,289 @@
...
+        String result = "test-result";
+        when(mockConverter.convert(cassandraRow)).thenReturn(result);
+        doThrow(new RuntimeException("Output failed")).when(mockOutput).collect(result);
+
+        assertThatThrownBy(() -> emitter.emitRecord(cassandraRow, mockOutput, mockSplit))
+                .isInstanceOf(RuntimeException.class)
+                .hasMessage("Failed to deserialize Cassandra row");
+
+        verify(mockConverter).convert(cassandraRow);
+        verify(mockOutput).collect(result);
```
[the remark this answers is not in the document]
Comment (Poorvankbhatia) [reply flink-connector-cassandra/src/test/java/org/apache/flink/connector/cassandra/source/reader/CassandraRowEmitterTest.java]: Corrected. The new test case looks like this:

```
@Test
    void testEmitRecordWithOutputException() {
        // Setup - converter succeeds but output.collect() fails
        String result = "test-result";
        when(mockConverter.convert(cassandraRow)).thenReturn(result);
        RuntimeException outputException = new RuntimeException("Output failed");
        doThrow(outputException).when(mockOutput).collect(result);
        assertThatThrownBy(() -> emitter.emitRecord(cassandraRow, mockOutput, mockSplit))
                .isInstanceOf(RuntimeException.class)
                .hasMessage("Failed to deserialize Cassandra row")
                .hasCause(outputException);
        verify(mockConverter).convert(cassandraRow);
    }
```

File: flink-connector-cassandra/src/test/java/org/apache/flink/connector/cassandra/source/reader/converter/CassandraRowToPojoConverterTest.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (Poorvankbhatia) [reply flink-connector-cassandra/src/test/java/org/apache/flink/connector/cassandra/source/reader/converter/CassandraRowToPojoConverterTest.java:49]: For now i have removed, if required in the future it can be added.

File: flink-connector-cassandra/src/test/java/org/apache/flink/connector/cassandra/table/CassandraClusterBuilderTest.java [hunk 14]
```diff
@@ -0,0 +1,214 @@
...
+        assertThat(cluster.getConfiguration().getProtocolOptions().getPort()).isEqualTo(9042);
+        assertThat(cluster.getConfiguration().getQueryOptions().getConsistencyLevel())
+                .isEqualTo(ConsistencyLevel.LOCAL_ONE);
+
+        cluster.close();
+    }
+
+    /** Tests parameterized constructor with valid configuration values. */
+    @Test
+    void testParameterizedConstructor() {
```
[the remark this answers is not in the document]
Comment (Poorvankbhatia) [reply flink-connector-cassandra/src/test/java/org/apache/flink/connector/cassandra/table/CassandraClusterBuilderTest.java:50]: Added

File: flink-connector-cassandra/src/test/java/org/apache/flink/connector/cassandra/table/CassandraDynamicTableSourceITCase.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (Poorvankbhatia) [reply flink-connector-cassandra/src/test/java/org/apache/flink/connector/cassandra/table/CassandraDynamicTableSourceITCase.java]: Added.

File: flink-connector-cassandra/src/test/java/org/apache/flink/connector/cassandra/table/CassandraTableFactoryTest.java [hunk 15]
```diff
@@ -0,0 +1,238 @@
...
+import org.junit.jupiter.api.Test;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+/** Unit tests for {@link CassandraTableFactory}. */
+class CassandraTableFactoryTest {
```
[the remark this answers is not in the document]
Comment (Poorvankbhatia) [reply flink-connector-cassandra/src/test/java/org/apache/flink/connector/cassandra/table/CassandraTableFactoryTest.java:41]: Added these methods.

File: flink-connector-cassandra/src/test/java/org/apache/flink/connector/cassandra/table/mapper/CassandraFieldMapperFactoryTest.java [hunk 16]
```diff
@@ -0,0 +1,300 @@
...
+import org.apache.flink.table.types.logical.YearMonthIntervalType;
+import org.apache.flink.table.types.logical.ZonedTimestampType;
+
+import org.junit.jupiter.api.Test;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.Assert.assertThrows;
+
+/** Unit tests for {@link CassandraFieldMapperFactory}. */
+class CassandraFieldMapperFactoryTest {
```
[the remark this answers is not in the document]
Comment (Poorvankbhatia) [reply flink-connector-cassandra/src/test/java/org/apache/flink/connector/cassandra/table/mapper/CassandraFieldMapperFactoryTest.java:59]: Correced it. PTAL. All test methods follow the pattern:
  assertThat(mapper).isNotNull().isInstanceOf(SomeClass.class);

File: flink-connector-cassandra/src/test/java/org/apache/flink/connector/cassandra/table/mapper/CollectionFieldMappersTest.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (Poorvankbhatia) [reply flink-connector-cassandra/src/test/java/org/apache/flink/connector/cassandra/table/mapper/CollectionFieldMappersTest.java:48]: All test methods are already using mapper.extractRowFieldValue(mockRow, "field"). PTAL, if this looks good?

File: flink-connector-cassandra/src/test/java/org/apache/flink/connector/cassandra/table/CassandraClusterBuilderTest.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (echauchot) [reply flink-connector-cassandra/src/test/java/org/apache/flink/connector/cassandra/table/CassandraClusterBuilderTest.java:50]: thx

File: flink-connector-cassandra/src/test/java/org/apache/flink/connector/cassandra/table/CassandraDynamicTableSourceITCase.java [hunk 17]
```diff
@@ -0,0 +1,1520 @@
...
+        assertThat(nestedTuple.getField(2)).isNotNull();
+        /** map */
+
+        /** Test collections of tuples */
+        assertThat(row.getField(4)).isNotNull();
+        /** list_of_tuples */
+        assertThat(row.getField(5)).isNotNull();
+        /** map_with_tuple_key */
+        assertThat(row.getField(6)).isNotNull();
+        /** map_with_tuple_value */
```
[the remark this answers is not in the document]
Comment (Poorvankbhatia) [reply flink-connector-cassandra/src/test/java/org/apache/flink/connector/cassandra/table/CassandraDynamicTableSourceITCase.java]: Added

[the remark this answers is not in the document]
Comment (echauchot) [reply flink-connector-cassandra/src/test/java/org/apache/flink/connector/cassandra/table/CassandraDynamicTableSourceITCase.java]: thx

File: flink-connector-cassandra/src/test/java/org/apache/flink/connector/cassandra/table/CassandraDynamicTableSourceITCase.java [hunk 18]
```diff
@@ -0,0 +1,1520 @@
...
+        assertThat(row.getField(8)).isNotNull();
+        /** single_map */
+
+        /** Test nested empty collections */
+        Object[] listOfEmptyLists = (Object[]) row.getField(9);
+        assertThat(listOfEmptyLists).hasSize(2);
+        /** First list is empty (null), second has one item */
+        assertThat((Object[]) listOfEmptyLists[1]).containsExactly("item");
+
+        assertThat(row.getField(10)).isNotNull();
```
[the remark this answers is not in the document]
Comment (echauchot) [reply flink-connector-cassandra/src/test/java/org/apache/flink/connector/cassandra/table/CassandraDynamicTableSourceITCase.java]: thx

File: flink-connector-cassandra/src/test/java/org/apache/flink/connector/cassandra/table/CassandraDynamicTableSourceITCase.java [hunk 19]
```diff
@@ -0,0 +1,1520 @@
...
+                        + "  list_with_values ARRAY<STRING>,"
+                        + "  map_with_values MAP<STRING, STRING>,"
+                        + "  single_list ARRAY<DOUBLE>,"
+                        + "  single_set MULTISET<STRING>,"
+                        + "  single_map MAP<INT, STRING>,"
+                        + "  list_of_empty_lists ARRAY<ARRAY<STRING>>,"
+                        + "  map_of_empty_maps MAP<STRING, MAP<STRING, INT>>,"
+                        + "  large_blob VARBINARY,"
+                        + "  tuple_with_nulls ROW<field1 STRING, field2 INT, field3 BOOLEAN>,"
+                        + "  partial_udt ROW<street STRING, city STRING, zipcode INT, country STRING>"
```
[the remark this answers is not in the document]
Comment (echauchot) [reply flink-connector-cassandra/src/test/java/org/apache/flink/connector/cassandra/table/CassandraDynamicTableSourceITCase.java]: thx

File: flink-connector-cassandra/src/test/java/org/apache/flink/connector/cassandra/table/CassandraDynamicTableSourceITCase.java [hunk 20]
```diff
@@ -0,0 +1,1600 @@
...
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+/**
+ * Comprehensive integration tests for Cassandra Dynamic Table Source covering all data types,
+ * complex scenarios, edge cases, and error conditions. Tests both basic functionality and
+ * real-world complex data structures to ensure complete mapper coverage.
+ */
+@ExtendWith(MiniClusterExtension.class)
+class CassandraDynamicTableSourceITCase {
```
[the remark this answers is not in the document]
Comment (echauchot) [reply flink-connector-cassandra/src/test/java/org/apache/flink/connector/cassandra/table/CassandraDynamicTableSourceITCase.java:46]: thx

File: flink-connector-cassandra/src/test/java/org/apache/flink/connector/cassandra/table/CassandraTableFactoryTest.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (echauchot) [reply flink-connector-cassandra/src/test/java/org/apache/flink/connector/cassandra/table/CassandraTableFactoryTest.java:46]: thx

**Review by echauchot — APPROVED** [review]
Thanks for addressing the review comments @Poorvankbhatia and thank for this very good work ! LGTM

**Comment by Poorvankbhatia** [comment]
Thanks @echauchot . Can you re run the CI again. (I had missed the license/notice in the shaded jar, so I corrected it with the last commit.) And should i stash now, if you are done with the review?

**Comment by echauchot** [comment]
> Thanks @echauchot . Can you re run the CI again. (I had missed the license/notice in the shaded jar, so I corrected it with the last commit.) And should i stash now, if you are done with the review?

Done. 
Yes please squash to the first commit. This commit granularity is fine. For later changes to parts of the code we will keep the commit granularity smaller for isolation purposes, but for new feature addition 1 commit is good.
